### PR TITLE
Add 4.12 heterogeneous on amd64 cluster upgrade to 4.13

### PR DESCRIPTION
**Pull request by shellyyang1989**
Refers [MIXEDARCH-135](https://issues.redhat.com/browse/MIXEDARCH-135)

### CODE REVIEW

**Comment by shellyyang1989** [comment]
cc @aleskandro @jianlinliu  @liangxia @jhou1 @YuLi517 for review. Thanks!

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.13__multi-nightly-4.13-upgrade-from-multi-stable-4.12.yaml [hunk 1]
```diff
@@ -0,0 +1,155 @@
+base_images:
```
Comment (liangxia) [inline ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.13__multi-nightly-4.13-upgrade-from-multi-stable-4.12.yaml:1]: The file rename is unexpected.

Comment (shellyyang1989) [reply ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.13__multi-nightly-4.13-upgrade-from-multi-stable-4.12.yaml:1]: Hmm, I renamed it as the file will only hold the multi-arch to multi-arch upgrade. Any side effect if it's renamed?

Comment (liangxia) [reply ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.13__multi-nightly-4.13-upgrade-from-multi-stable-4.12.yaml:1]: Originally, it contains only multi-arch upgrade test in it, with the same name style as the others.

Comment (liangxia) [reply ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.13__multi-nightly-4.13-upgrade-from-multi-stable-4.12.yaml:1]: Also i think we do not support cross arch upgrade at the moment, e.g, `amd64-nightly-4.13-upgrade-from-arm64-stable-4.12.yaml`, so there is no need to duplicate the arch info twice in the file name (thus in the job name, which makes the job name longer).

Comment (shellyyang1989) [reply ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.13__multi-nightly-4.13-upgrade-from-multi-stable-4.12.yaml:1]: Makes sense. Thanks!

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.13__multi-nightly-4.13-upgrade-from-stable-4.12.yaml [hunk 2]
```diff
@@ -99,6 +104,50 @@ tests:
...
+      OCP_ARCH: arm64
+      UPGRADE_CHECK_RUN_TAGS: '@upgrade-check and @4.13 and @aws-ipi and @heterogeneous
+        and @network-ovnkubernetes and not @fips'
+      UPGRADE_PRE_RUN_TAGS: '@upgrade-prepare and @4.12 and @aws-ipi and @heterogeneous
+        and @network-ovnkubernetes and not @fips'
+    test:
+    - chain: openshift-upgrade-qe-test
+    - ref: ipi-install-heterogeneous
+    workflow: cucushift-installer-rehearse-aws-ipi-ovn-ipsec
+- as: aws-ipi-ovn-ipsec-x86-p2-f14
```
Comment (aleskandro) [inline ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.13__multi-nightly-4.13-upgrade-from-stable-4.12.yaml:123]: Let's stay consistent with the way we print the architecture: (arm64, amd64) or (aarch64, x86_64)

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.13__multi-nightly-4.13-upgrade-from-stable-4.12.yaml [hunk 3]
```diff
@@ -99,6 +104,50 @@ tests:
...
+      BASE_DOMAIN: qe.devcluster.openshift.com
+      UPGRADE_CHECK_RUN_TAGS: '@upgrade-check and @4.13 and @aws-ipi and @heterogeneous
+        and @network-ovnkubernetes and not @fips'
+      UPGRADE_PRE_RUN_TAGS: '@upgrade-prepare and @4.12 and @aws-ipi and @heterogeneous
+        and @network-ovnkubernetes and not @fips'
+    test:
+    - chain: openshift-upgrade-qe-test
+    - ref: ipi-install-heterogeneous
+    workflow: cucushift-installer-rehearse-aws-ipi-ovn-ipsec
+- as: aws-ipi-x86-p2-f14
```
Comment (aleskandro) [inline ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.13__multi-nightly-4.13-upgrade-from-stable-4.12.yaml:137]: same as above, I suggest amd64 instead of x86 here.

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.13__multi-nightly-4.13-upgrade-from-stable-4.12.yaml [hunk 4]
```diff
@@ -83,12 +74,25 @@ tests:
     - ref: ipi-install-heterogeneous
     - chain: openshift-upgrade-qe-test
     workflow: cucushift-installer-rehearse-aws-ipi
+- as: aws-ipi-arm64-p2-f14
+  cron: 9 14 5,20 * *
+  steps:
+    cluster_profile: aws-qe
+    env:
+      ADDITIONAL_WORKER_ARCHITECTURE: x86_64
```
Comment (shellyyang1989) [inline ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.13__multi-nightly-4.13-upgrade-from-stable-4.12.yaml:82]: @aleskandro Could you please help confirm if it's needed and if the value is correct? Thank you!

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.13__multi-nightly-4.13-upgrade-from-stable-4.12.yaml (hunk 3 above)
Comment (shellyyang1989) [reply ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.13__multi-nightly-4.13-upgrade-from-stable-4.12.yaml]: Updated. Thanks!

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.13__multi-nightly-4.13-upgrade-from-stable-4.12.yaml (hunk 4 above)
Comment (aleskandro) [reply ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.13__multi-nightly-4.13-upgrade-from-stable-4.12.yaml:82]: Yes, I'd keep it explicitly set and when we start from arm64 clusters, it's correct to set additional workers with x86_64 arch.

Comment (aleskandro) [reply ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.13__multi-nightly-4.13-upgrade-from-stable-4.12.yaml:82]: You also need to set

`ADDITIONAL_WORKER_VM_TYPE: m5.xlarge` (is this the default one we're using for x86 machines right?)

Comment (shellyyang1989) [reply ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.13__multi-nightly-4.13-upgrade-from-stable-4.12.yaml:82]: It's m6i.xlarge by default in 4.13. But m5.xlarge is supported as well.

Comment (aleskandro) [reply ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.13__multi-nightly-4.13-upgrade-from-stable-4.12.yaml:82]: Ok to use m6i.xlarge as well

**Comment by shellyyang1989** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.13-multi-nightly-4.13-upgrade-from-stable-4.12-aws-ipi-arm64-p2-f14

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.13__multi-nightly-4.13-upgrade-from-stable-4.12.yaml (hunk 4 above)
Comment (shellyyang1989) [reply ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.13__multi-nightly-4.13-upgrade-from-stable-4.12.yaml:82]: Updated. Thanks!

**Comment by aleskandro** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.13-multi-nightly-4.13-upgrade-from-stable-4.12-aws-ipi-arm64-p2-f14

**Comment by jianlinliu** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.13-multi-nightly-4.13-upgrade-from-stable-4.12-aws-ipi-arm64-p2-f14

**Comment by jianlinliu** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.13-multi-nightly-4.13-upgrade-from-stable-4.12-aws-ipi-arm64-p2-f14

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.13__multi-nightly-4.13-upgrade-from-stable-4.12.yaml [hunk 5]
```diff
@@ -83,12 +74,26 @@ tests:
...
+      ADDITIONAL_WORKER_VM_TYPE: m6i.xlarge
+      BASE_DOMAIN: qe.devcluster.openshift.com
+      OCP_ARCH: arm64
```
Comment (jianlinliu) [inline ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.13__multi-nightly-4.13-upgrade-from-stable-4.12.yaml:85]: Per the error message in https://qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gs/qe-private-deck/pr-logs/pull/openshift_release/36173/rehearse-36173-periodic-ci-openshift-openshift-tests-private-release-4.13-multi-nightly-4.13-upgrade-from-stable-4.12-aws-ipi-arm64-p2-f14/1625679162758926336:
```
level=error msg=Error: creating EC2 Instance: InvalidParameterValue: The architecture 'x86_64' of the specified instance type does not match the architecture 'arm64' of the specified AMI. Specify an instance type and an AMI that have matching architectures, and try again. You can use 'describe-instance-types' or 'describe-images' to discover the architecture of the instance type or AMI.
```

`COMPUTE_NODE_TYPE` and `CONTROL_PLANE_INSTANCE_TYPE` need to be overwritten, because by default, `cucushift-installer-rehearse-aws-ipi-provision` set them to x86 instance type.

Comment (shellyyang1989) [reply ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.13__multi-nightly-4.13-upgrade-from-stable-4.12.yaml:85]: ah, thank you Jianlin for the debug!!

**Comment by shellyyang1989** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.13-multi-nightly-4.13-upgrade-from-stable-4.12-aws-ipi-arm64-p2-f14

**Comment by aleskandro** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.13-multi-nightly-4.13-upgrade-from-stable-4.12-aws-ipi-arm64-p2-f14

**Comment by shellyyang1989** [comment]
```
level=error msg=The bootstrap machine failed to download the release image
level=info msg=time="2023-02-15T18:31:07Z" level=warning msg="image platform ({amd64 linux  [] }) does not match the expected platform ({arm64 linux  [] })"
level=info msg=208e8b82f4da3c35863bcdaabdf311e58c32841adca7ee6307aca658811fd7c8
level=info msg=ERROR: release image arch amd64 does not match host arch arm64
```
[36mINFO[0m[2023-02-15T18:21:35Z] Resolved release latest to quay.io/openshift-release-dev/ocp-release:4.12.4-multi

Looks like multi-arch payload is unable to install single-arch cluster. @aleskandro

**Comment by aleskandro** [comment]
Hi @shellyyang1989 which one is failing?

> ```
> level=error msg=The bootstrap machine failed to download the release image
> level=info msg=time="2023-02-15T18:31:07Z" level=warning msg="image platform ({amd64 linux  [] }) does not match the expected platform ({arm64 linux  [] })"
> level=info msg=208e8b82f4da3c35863bcdaabdf311e58c32841adca7ee6307aca658811fd7c8
> level=info msg=ERROR: release image arch amd64 does not match host arch arm64
> ```
> 
> �[36mINFO�[0m[2023-02-15T18:21:35Z] Resolved release latest to quay.io/openshift-release-dev/ocp-release:4.12.4-multi �[36mINFO�[0m[2023-02-15T18:23:21Z] Imported release 4.12.4 created at 2023-02-15 09:46:09 +0000 UTC with 183 images to tag release:latest
> 
> Is the payload multi-arch?

Yeah, it seems we cannot because the internal registry still imports only the single arch manifest from the multi payload manifest-list and since this job is running in an amd64 cluster, the imported manifest used at bootstrap time is for an amd64 payload. Let me follow up on arm64->multi paths in Prow later, and let's test just amd64->multi for now.

**Comment by aleskandro** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.13-multi-nightly-4.13-upgrade-from-stable-4.12-aws-ipi-ovn-ipsec-amd64-p2-f14

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.13__multi-nightly-4.13-upgrade-from-stable-4.12.yaml [hunk 6]
```diff
@@ -83,12 +74,28 @@ tests:
     - ref: ipi-install-heterogeneous
     - chain: openshift-upgrade-qe-test
     workflow: cucushift-installer-rehearse-aws-ipi
+- as: aws-ipi-arm64-p2-f14
+  cron: 9 14 5,20 * *
```
Comment (deepsm007) [inline ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.13__multi-nightly-4.13-upgrade-from-stable-4.12.yaml:78]: Add `cluster: arm01` to trigger the same job arm64 build cluster

Comment (shellyyang1989) [reply ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.13__multi-nightly-4.13-upgrade-from-stable-4.12.yaml:78]: Thank you Deep!!

**Comment by aleskandro** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.13-multi-nightly-4.13-upgrade-from-stable-4.12-aws-ipi-arm64-p2-f14

**Comment by shellyyang1989** [comment]
```
[2023-02-16T15:33:47Z] 
  * could not resolve inputs: could not determine inputs for step [input:dev-scripts]: could not resolve base image from ocp-kni-arm64/test:dev-scripts: imagestreamtags.image.openshift.io "test" not found 
```
Test image is not found.

@aleskandro wonder why arm tests don't set the `cluster: arm01` but work well.

**Comment by aleskandro** [comment]
> ```
> [2023-02-16T15:33:47Z] 
>   * could not resolve inputs: could not determine inputs for step [input:dev-scripts]: could not resolve base image from ocp-kni-arm64/test:dev-scripts: imagestreamtags.image.openshift.io "test" not found 
> ```
> 
> Test image is not found.
> 
> @aleskandro wonder why arm tests don't set the `cluster: arm01` but work well.

TLDR: This issue does not affect customer installations and will go away when we upgrade the Prow clusters to 4.13.

The arm01 cluster is an arm64 prow cluster, useful when you need to run some steps/jobs in arm64 workers. It's not required for running arm64 installations.

However, we still miss some CI images enablement in arm01, as in the error you got. I suggest **removing the scenarios starting from arm64 and having it as a follow up (we need to ensure the required images for the tests are built in arm01 first)** (we can look at them from an installation automation perspective first).

---

The problem is that we're running tests from prow clusters based on 4.12.z, and the 4.12 internal registry doesn't support manifest list images the way we use them in Prow.

The arm builds work by using a single-arch manifest that is pulled at the beginning of the tests. When a host pulls an Image Manifest (let's say a single-arch image), it doesn't look at the architecture of the image and it does what a user asks (even try executing it)...

So, in single-arch arm64 tests we (1) resolve the payload, (2) mirror it in the internal registry, (3) use the amd64 images (installer, test etc..) to run the installation and the execution of the e2e cases.

Instead, a multi-arch payload is based on v2 ManifestList images (lets' say multi-arch images). In this case, an host pulling the ManifestList, will (1) look the manifestlist, (2) get the Manifest for its architecture, (3) pull the (single-arch) manifest, and (4) use the amd64 images (installer, test...) to run the installation.

This is what happens in the internal registry until OCP 4.12. We mirror the multi-payload and get the amd64 image as the runner nodes are amd64. Then, we create an arm64 bootstrap node, and it executes this code: in https://github.com/openshift/installer/blob/4daff6117fce1d73774b02caf605d9e1b836dd7c/data/data/bootstrap/files/usr/local/bin/release-image-download.sh.template#L36-L42. Since the bootstrap host is now pulling from the Manifest extracted from the ManifestList, it gets the amd64 image, and fails the validation.

@deepsm007 suggested using the arm01 cluster so that the internal registry would extract the arm64 manifest from the manifest list.

Finally, although the bootstrap host in those lines sees a single-arch image, that image carries the list of multi-arch manifest list payload components as well. So in the case of amd64 control plane (and bootstrap node), it works fine just because that validation passes, and then the multi-arch, manifest list, images are used as refs by the OCP components being installed.

**Comment by shellyyang1989** [comment]
Thank you Ales for sharing and explanation! Removed arm tests.

**Comment by shellyyang1989** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.13-multi-nightly-4.13-upgrade-from-stable-4.12-aws-ipi-ovn-ipsec-amd64-p2-f14

**Comment by shellyyang1989** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.13-multi-nightly-4.13-upgrade-from-stable-4.12-aws-ipi-ovn-ipsec-amd64-p2-f14

**Comment by shellyyang1989** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.13-multi-nightly-4.13-upgrade-from-stable-4.12-aws-ipi-ovn-ipsec-amd64-p2-f14

**Comment by shellyyang1989** [comment]
Upgrade and scale up steps passed.

```
INFO[2023-02-19T09:49:47Z] Running step aws-ipi-ovn-ipsec-amd64-p2-f14-cucushift-upgrade-toimage. 
INFO[2023-02-19T11:08:17Z] Step aws-ipi-ovn-ipsec-amd64-p2-f14-cucushift-upgrade-toimage succeeded after 1h18m30s. 
INFO[2023-02-19T11:08:17Z] Running step aws-ipi-ovn-ipsec-amd64-p2-f14-ipi-install-heterogeneous. 
INFO[2023-02-19T11:13:37Z] Step aws-ipi-ovn-ipsec-amd64-p2-f14-ipi-install-heterogeneous succeeded after 5m20s. 
```

**Comment by shellyyang1989** [comment]
@aleskandro @liangxia It's ready for merge. Thanks!

File: ci-operator/step-registry/openshift/upgrade/qe/test/heterogeneous/openshift-upgrade-qe-test-heterogeneous-chain.metadata.json [hunk 7]
```diff
@@ -0,0 +1,13 @@
+{
+	"path": "openshift/upgrade/qe/test/heterogeneous/openshift-upgrade-qe-test-heterogeneous-chain.yaml",
+	"owners": {
+		"approvers": [
+			"jhou1",
+			"liangxia",
+			"jianzhangbjz",
+			"kuiwang02",
```
[the remark this answers is not in the document]
Comment (shellyyang1989) [reply ci-operator/step-registry/openshift/upgrade/qe/test/heterogeneous/openshift-upgrade-qe-test-heterogeneous-chain.metadata.json:8]: Updated! Please take a look @aleskandro . Thank you!

[the remark this answers is not in the document]
Comment (aleskandro) [reply ci-operator/step-registry/openshift/upgrade/qe/test/heterogeneous/openshift-upgrade-qe-test-heterogeneous-chain.metadata.json:8]: Thanks
